core/arch/arch.c

```diff
@@ -3254,7 +3254,10 @@ check_syscall_method(dcontext_t *dcontext, instr_t *instr)
         ASSERT(get_syscall_method() == SYSCALL_METHOD_UNINITIALIZED ||
                get_syscall_method() == SYSCALL_METHOD_INT);
 #    ifdef LINUX
-        if (new_method == SYSCALL_METHOD_SYSENTER) {
+        if (new_method ==
+            SYSCALL_METHOD_SYSENTER IF_X86_32(||
+                                              (new_method == SYSCALL_METHOD_SYSCALL &&
+                                               cpu_info.vendor == VENDOR_AMD))) {
```

Comment:
Please update other places that assume syscall behaves "normally": like does_syscall_ret_to_callsite().  Not sure if there are others.

Do DR's own syscalls, both the static ones in drlibc_x86.asm (are they still `int 0x80` here?) and generated (is it again using `int 0x80` and never `syscall`??  Please double-check) work??  I could imagine lots of problems with the weird `syscall` behavior.

Reply:
> Please update other places that assume syscall behaves "normally": like does_syscall_ret_to_callsite(). Not sure if there are others.

I am sorry but do not understand, what should update and where?

> Do DR's own syscalls, both the static ones in drlibc_x86.asm 

In drlibc_x86.asm file, there is no case for 32-bit where used `syscall` instruction.
Here is a fragment from that static code.
![Screenshot from 2021-09-18 00-17-38](https://user-images.githubusercontent.com/18750457/133849109-047ccc76-f392-4fd3-9eb9-943fb0a3c761.png)

Reply:
> > Please update other places that assume syscall behaves "normally": like does_syscall_ret_to_callsite(). Not sure if there are others.
> 
> I am sorry but do not understand, what should update and where?

As mentioned, the `does_syscall_ret_to_callsite()` function needs updating, since this `syscall` certainly does not return.

Reply:
Updated